File: src/main/java/io/github/tanguygab/conditionalactions/actions/Action.java
package io.github.tanguygab.conditionalactions.actions;

import io.github.tanguygab.conditionalactions.ConditionalActions;
import lombok.Getter;
import lombok.experimental.Accessors;
import me.clip.placeholderapi.PlaceholderAPI;
import org.bukkit.OfflinePlayer;
import org.intellij.lang.annotations.Language;

import java.util.regex.Pattern;

@Getter
public abstract class Action {

    private final Pattern pattern;
    @Accessors(fluent = true) private final boolean replaceMatch;

    public Action(@Language("RegExp") String regex, boolean replaceMatch) {
        this.pattern = Pattern.compile(regex);
        this.replaceMatch = replaceMatch;
    }

    public Action(@Language("RegExp") String regex) {
        this(regex,true);
    }

    public abstract String getSuggestion();
    public abstract void execute(OfflinePlayer player, String match);

    protected ConditionalActions getPlugin() {
        return ConditionalActions.getInstance();
    }
    protected void sync(Runnable run) {
        getPlugin().sync(run);
    }
    protected String parsePlaceholders(OfflinePlayer player, String string) {
        return PlaceholderAPI.setPlaceholders(player,string);
    }
    protected int parseInt(String str, int def) {
        try {return Integer.parseInt(str);}
        catch (Exception e) {return def;}
    }

    protected String getArgumentSeparator() {
        return getPlugin().getActionManager().getArgumentSeparator();
    }
    protected String[] split(String args) {
        return args.split(Pattern.quote(getArgumentSeparator()));
    }
    protected String getSuggestionWithArgs(String... args) {
        return String.join(getArgumentSeparator(),args);
    }
}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/actions/Action.kt
package io.github.tanguygab.conditionalactions.actions

import io.github.tanguygab.conditionalactions.ConditionalActions
import io.papermc.paper.threadedregions.scheduler.ScheduledTask
import me.clip.placeholderapi.PlaceholderAPI
import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.features.PlaceholderManagerImpl
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.concurrent.CompletableFuture

abstract class Action(val pattern: Regex, val replaceMatch: Boolean = true) {
    protected val plugin = ConditionalActions.INSTANCE
    protected val mm get() = MiniMessage.miniMessage()

    open val description = mm.deserialize("<dark_red>No description.")
    abstract fun getSuggestion(): String

    abstract fun execute(player: OfflinePlayer?, match: String)

    protected fun parsePlaceholders(player: OfflinePlayer?, string: String): String {
        if (plugin.server.pluginManager.isPluginEnabled("TAB") && player?.player != null) {
            val tabPlayer = TAB.getInstance().getPlayer(player.uniqueId)
            val placeholders = PlaceholderManagerImpl.detectPlaceholders(string)
                .map { TAB.getInstance().placeholderManager.getPlaceholder(it) }

            var string = string
            for (placeholder in placeholders) string = string.replace(placeholder.identifier, placeholder.parse(tabPlayer))
            return string
        }
        return PlaceholderAPI.setPlaceholders(player, string)
    }

    protected fun sync(player: Player?, run: Runnable) {
        val future = CompletableFuture<Void>()
        val task: (ScheduledTask) -> Unit = {
            run.run()
            future.complete(null)
        }
        if (player == null) plugin.server.globalRegionScheduler.run(plugin, task)
        else player.scheduler.run(plugin, task, null)
        future.get()
    }

    protected fun parseInt(str: String, def: Int) = str.toIntOrNull() ?: def

    protected fun split(args: String): List<String> = args.split(plugin.actionManager.argumentSeparator)

    protected fun getSuggestionWithArgs(vararg args: String?) = args.joinToString(plugin.actionManager.argumentSeparator)
}